File: c/entrada/entrada.c
#include "../entrada/entrada.h"

struct config_entrada InicialziarConfig(void) {
  struct config_entrada nova_config;

  nova_config.tecla_confirmacao = ENTER;
  nova_config.tecla_avancar = SETA_BAIXO;
  nova_config.tecla_voltar = SETA_CIMA;
  nova_config.num_items = 0;
  nova_config.opcao_atual = OPCAO_INICIAL_PADRAO;
    
  return nova_config;
} 

int SelecionarOpcao(struct config_entrada *config) {

  int opcao_atual = config->opcao_atual;
  int num_items   = config->num_items;

  char buffer[3] = { 0 };
  read(STDIN_FILENO, buffer, 3);

  int tecla = (int)buffer[0] + (int)buffer[1] + (int)buffer[2];

  switch (tecla) {

    case ENTER:
    case ESPACO:
      if (config->tecla_confirmacao == tecla)
        return config->opcao_atual;
      else
        return -1;
    
    case SETA_CIMA:
    case SETA_ESQUERDA:
      if (config->tecla_voltar == tecla) {
        if (opcao_atual > 0)
          config->opcao_atual -= 1;
        else
          config->opcao_atual = config->num_items;
      }
    
      return -1;
    
    case SETA_BAIXO:
    case SETA_DIREITA:
      if (config->tecla_avancar == tecla) {
        if (opcao_atual < num_items)
          config->opcao_atual += 1;
        else
          config->opcao_atual = 0;
      }
      return -1;

    case ESC:
      return -2;
    
    default:
      return -4;
  }
}

File: pascal/entrada.pas
unit entrada;

interface

const
	ESC = 27;
    SETA_CIMA = 183;
    SETA_BAIXO = 184;
    SETA_DIREITA = 185;
    SETA_ESQUERDA = 186;
    ENTER = 10;
    ESPACO = 32;

type
	ConfigEntrada = record
		teclaConfirmacao: integer;
		teclaAvancar: integer;
		teclaVoltar: integer;
		numItems: integer;
		opcaoAtual: integer
	end;

function inicializarConfig(): ConfigEntrada;
function selecionarOpcao(var config: ConfigEntrada): integer;

implementation

function inicializarConfig(): ConfigEntrada;

begin
	inicializarConfig.teclaConfirmacao := ENTER;
	inicializarConfig.teclaAvancar := SETA_BAIXO;
	inicializarConfig.teclaVoltar := SETA_CIMA;
	inicializarConfig.numItems := 0;
	inicializarConfig.opcaoAtual := 0; // Opção Inicial Padrão
end;

function selecionarOpcao(var config: ConfigEntrada): integer;

var
	tecla: integer;

begin
	read(tecla);

	case tecla of
		ENTER, ESPACO: 
		begin
			if config.teclaConfirmacao = tecla then
				selecionarOpcao := config.opcaoAtual
			else
				selecionarOpcao := -1;
		end;

		SETA_CIMA, SETA_ESQUERDA:
		begin
			if config.teclaVoltar = tecla then
			begin
				if config.opcaoAtual > 0 then
					config.opcaoAtual -= 1
				else
					config.opcaoAtual := config.numItems;
			end;

			selecionarOpcao := -1;
		end;

		SETA_BAIXO, SETA_DIREITA:
		begin
			if config.teclaAvancar = tecla then
			begin
				if config.opcaoAtual < config.numItems then
					config.opcaoAtual += 1
				else
					config.opcaoAtual := 0;
			end;

			selecionarOpcao := -1;
		end;

		ESC:
			selecionarOpcao := -2;
		else
			selecionarOpcao := -4;
	end;
end;

end.
